src/main/kotlin/com/petqua/application/order/OrderService.kt

```diff
@@ -23,90 +33,80 @@ import com.petqua.exception.order.ShippingAddressException
-
-        // TODO 상품 유효성 검증 - 올바른 옵션 매칭인가?
-        val productOptions = productOptionRepository.findByProductIdIn(productIds)
-
-        command.orderProductCommands.forEach { productOptionCommand ->
-            productOptions.find { it.productId == productOptionCommand.productId }?.let {
-                throwExceptionWhen(!it.isSame(productOptionCommand.toProductOption())) {
-                    ProductException(INVALID_PRODUCT_OPTION)
-                }
-            } ?: throw ProductException(INVALID_PRODUCT_OPTION)
-        }
-
-        // TODO 배송지 존재 검증
         val shippingAddress = shippingAddressRepository.findByIdOrThrow(command.shippingAddressId) {
             ShippingAddressException(NOT_FOUND_SHIPPING_ADDRESS)
         }
+        val productSnapshots =
+            productSnapshotRepository.findAllByProductIdIn(orderProducts.productIds).associateBy { it.productId }
+        validateProductSnapshots(orderProducts, productSnapshots)
 
-        // TODO 총 가격 검증
-        // 1. 상품 가격
-        command.orderProductCommands.forEach { productCommand ->
-            val product = productById[productCommand.productId]
-                ?: throw OrderException(PRODUCT_NOT_FOUND)
-            val productOption = productOptions.find { it.productId == product.id }
-                ?: throw ProductException(INVALID_PRODUCT_OPTION)
-
-            throwExceptionWhen(
-                productCommand.orderPrice.setScale(2) != (product.discountPrice + productOption.additionalPrice) * productCommand.quantity.toBigDecimal()
-                        || productCommand.deliveryFee.setScale(2) != product.getDeliveryFee(productCommand.deliveryMethod)
-            ) {
-                OrderException(
-                    ORDER_PRICE_NOT_MATCH
-                )
-            }
-        }
+        // TODO: TODO 재고 검증
+        val orders = saveOrders(command, productSnapshots, shippingAddress)
+        orderPaymentRepository.saveAll(orders.map { OrderPayment.from(it) })
+        return SaveOrderResponse(
+            orderId = orders.first().orderNumber.value,
+            orderName = orders.first().orderName.value,
+            successUrl = paymentGatewayClient.successUrl(),
+            failUrl = paymentGatewayClient.failUrl(),
+        )
+    }
 
-        // 3. 총 배송비 검증 (스토어로 묶인 뒤 배송비 검증)
-        val groupBy = products.groupBy { product ->
-            Pair(
-                product.storeId,
-                command.orderProductCommands.find { it.productId == product.id }?.deliveryMethod
-                    ?: throw OrderException(PRODUCT_NOT_FOUND)
-            )
-        }
-        val orderDeliveryFee = groupBy.map { (storeDeliveryMethod, products) ->
-            val deliveryMethod = storeDeliveryMethod.second
-            products.first().getDeliveryFee(deliveryMethod).toInt()
-        }.sum()
+    private fun getOrderProductsFrom(command: SaveOrderCommand): OrderProducts {
```

Comment:
전체적으로 `SaveOrderCommand`를 파라미터로 넘기면서 사용하고 있는데요.
`SaveOrderCommand`객체보다 내부 값을 넘기는 식으로 활용하는게 어떨까요.
`save()`처럼 public하게 사용하도록 열어둔 메서드에서는 관심사항이 캡슐화된 객체를 받는건 좋지만, 내부에서도 이 객체를 계속 사용하면 로직이 더 헷갈리는 것 같아요.

```suggestion
    private fun getOrderProductsFrom(orderProductIds: List<Long>): OrderProducts {
```